qmetaobject/src/qtdeclarative.rs

```diff
@@ -436,6 +436,79 @@ pub fn qml_register_type<T: QObject + Default + Sized>(
     }
 }
 
+cpp! {{
+    typedef QObject *(*QmlRegisterSingletonTypeCallback)(QQmlEngine *, QJSEngine *);
+}}
+
+/// Register a default-constructed object of specified type as a singleton QML object.
+///
+/// A new object will be created for each new instance of `QmlEngine`.
+///
+/// Refer to the Qt documentation for qmlRegisterSingletonType.
+pub fn qml_register_singleton_type<T: QObject + Sized + Default>(
+    uri: &std::ffi::CStr,
+    version_major: u32,
+    version_minor: u32,
+    type_name: &std::ffi::CStr,
+) {
+    let uri_ptr = uri.as_ptr();
+    let type_name_ptr = type_name.as_ptr();
+
+    extern "C" fn callback_fn<T: QObject + Default + Sized>(
+        _qml_engine: *mut c_void,
+        _js_engine: *mut c_void,
+    ) -> *mut c_void {
+        let obj: Box<RefCell<T>> = Box::new(RefCell::new(T::default()));
+        let obj_ptr = unsafe { T::cpp_construct(&obj) };
+        Box::into_raw(obj);
+        obj_ptr
+    };
+    let callback_fn: extern "C" fn(
+        _qml_engine: *mut c_void,
+        _js_engine: *mut c_void,
+    ) -> *mut c_void = callback_fn::<T>;
+
+    unsafe {
+        cpp!([uri_ptr as "const char*", version_major as "int", version_minor as "int",
+                type_name_ptr as "const char*", callback_fn as "QmlRegisterSingletonTypeCallback"] {
+            qmlRegisterSingletonType<QObject>(uri_ptr, version_major, version_minor, type_name_ptr,
```

Comment:
I think this registers the wrong QMetaObject (the one from QObject).
There is also a non-template overload that should be called instead.

Reply:
I was hoping to use the overload taking a C++ std::function so that I could accept a Rust closure instead of being forced to Default-construct the object.

But I guess I will have to change it back.

Reply:
The non-template overload can only be used to register a callback that returns a `QJSValue`.

`inline int qmlRegisterSingletonType(const char *uri, int versionMajor, int versionMinor, const char *typeName, QJSValue (*callback)(QQmlEngine *, QJSEngine *))`

Reply:
I am now calling functions from `qqmlprivate.h` directly. Sadly there seems to be no other way around the template type issue.